### DISCUSSION TITLE
Using ControlTemplate, DataTemplate, etc. only in C# code?

### DISCUSSION THREAD

**Discussion by idexus** [microsoft/microsoft-ui-xaml#8294] [General]
I'm working on a [library to fully support WinUI 3 in code](https://github.com/idexus/CodeMarkup-WinUI) using fluent methods, without having to use XAML. The only problem I have is with classes like DataTemplate, ControlTemplate, etc. that I can't create without this weird XAML workaround.

<details>
<summary> XAML workaround </summary>

```cs
namespace CodeMarkup.WinUI
{
    public delegate void TemplateBuildMethodWithParent<T>(T root, FrameworkElement parent) where T : FrameworkElement;
    public delegate void TemplateBuildMethod<T>(T root) where T : FrameworkElement;

    public abstract class FrameworkTemplate<TTemplate, TRoot>
        where TTemplate : FrameworkTemplate
        where TRoot : FrameworkElement
    {
        public static implicit operator TTemplate(FrameworkTemplate<TTemplate, TRoot> frameworkTemplate) => frameworkTemplate.xamlFrameworkTemplate;

        TTemplate xamlFrameworkTemplate;
        public FrameworkTemplate(TemplateBuildMethodWithParent<TRoot> build)
        {
            var guid = Guid.NewGuid().ToString();
            FrameworkTempateManager.HandlerMethodsWithParent[guid] = (root, parent) => build(root as TRoot, parent);
            CreateXamlFrameworkTemplate(guid);
        }

        public FrameworkTemplate(TemplateBuildMethod<TRoot> build)
        {
            var guid = Guid.NewGuid().ToString();
            FrameworkTempateManager.HandlerMethodsWithoutParent[guid] = root => build(root as TRoot);
            CreateXamlFrameworkTemplate(guid);
        }

        void CreateXamlFrameworkTemplate(string guid)
        {            
            var xamlControlTemplateString =
$@"<{typeof(TTemplate).Name}
    xmlns=""http://schemas.microsoft.com/winfx/2006/xaml/presentation""
    xmlns:root=""using:{typeof(TRoot).Namespace}""
    xmlns:manager=""using:{typeof(FrameworkTempateManager).Namespace}"">

    <root:{typeof(TRoot).Name} 
        manager:{typeof(FrameworkTempateManager).Name}.MethodId=""{guid}"" 
        manager:{typeof(FrameworkTempateManager).Name}.TemplatedParent=""{{Binding RelativeSource={{RelativeSource TemplatedParent}}}}""/>

</{typeof(TTemplate).Name}>";
            xamlFrameworkTemplate = (TTemplate)XamlReader.Load(xamlControlTemplateString);
        }
    }

    public class ControlTemplate<TRoot> : FrameworkTemplate<ControlTemplate, TRoot>
        where TRoot : FrameworkElement
    {
        public ControlTemplate(TemplateBuildMethodWithParent<TRoot> build) : base(build)
        {
        }
    }

    public class DataTemplate<TRoot> : FrameworkTemplate<DataTemplate, TRoot>
        where TRoot : FrameworkElement
    {
        public DataTemplate(TemplateBuildMethod<TRoot> build) : base(build)
        {
        }
    }
}
```
</details>

Isn't there some solution to create these objects only in code? 

e.g.

```cs
ControlTemplate itemTemplate = new ControlTemplate<Grid>((root, parent) =>
{
     // Create  content
});
```
and
```cs
DataTemplate itemTemplate = new DataTemplate<Grid>(root =>
{
    // Create  content
});
```

then using fluent API I can do something like this:

```cs
new ItemsRepeater()
    .ItemsSource(percentDataList)
    .Layout(new StackLayout {Orientation = Orientation.Vertical})
    .ItemTemplate(dataTemplate) 

// data template declaration (static field)

readonly static DataTemplate dataTemplate = new DataTemplate<Frame>(root =>
{
    root.Content = new Border()
        .Background(e => e.ResourceKey("SystemChromeLowColor").Source(root))
        .Margin(4)
        .Size(400, 20)
        .Child(
            new Rectangle()
                .HorizontalAlignment(HorizontalAlignment.Left)
                .Fill(e => e.ResourceKey("SystemAccentColor").Source(root))
                .Width(e => e.Path("Percent").Convert<double>(d => d * 4))
        );
});
```

**Comment by DarranRowe** [comment]
The best suggestion I have is to basically copy what the Xaml compiler does.
The Xaml compiler does output source files. For C#, it should output the generated .cs files in the obj directory.

**Reply by marcelwgn** [reply]
But doesn't the compiler also output XBF files for the XAML Stuff? I don't think there is "code-behind" for everything, some stuff really lives in XBF (e.g. Theme-ResourceDictionaries)

**Reply by DarranRowe** [reply]
The .xbf files are basically just "compressed" versions of the .xaml files. This means it is easy enough to infer how they work when going through the Application.LoadComponent call that directly uses the .xaml files.
For the rest of it, there is absolutely enough generated code to work everything out.

**Reply by marcelwgn** [reply]
Well yes, they are still somewhat XAML files if we don't want to be pendantic, however those are also copied to the bin folder so I think it's not too far fetched to say they are not entirely converted to code behind and don't exist during runtime.

The generated files seem to be only for hooking up events, bindings and generally all things interop with code behind. I don't think there is "absolutely enough to work everything out" but rather just something glueing your code together with the XAML world. However if you actually found out what happens when creating styles or datatemplates, feel free to point out where one can find that.

**Reply by idexus** [reply]
@DarranRowe I tried to go this way (Application.LoadComponent), however it's not that easy, a lot of things are internal. Besides, what is generated in cs files is not enough, the rest is in xbf files. Honestly, this approach is the next workaround. That's not what I'm about. Let's take a DataTemplate for example. It was enough to change the approach in the library, and give the possibility to implement the IDataTemplate interface using a virtual method, this would allow you to create a DataTemplate directly in the code. This is missing a lot.

from the library:

```cs
namespace Windows.UI.Xaml
{
    [ComImport]
    [ContractVersion(typeof(UniversalApiContract), 65536u)]
    [WebHostHidden]
    [ExclusiveTo(typeof(DataTemplate))]
    [Windows.Foundation.Metadata.Guid(2568007367u, 35509, 16664, 155, 198, 9, 244, 90, 53, 7, 61)]
    internal interface IDataTemplate
    {
        [MethodImpl(MethodImplOptions.InternalCall)]
        DependencyObject LoadContent();
    }
}
```

**Reply by DarranRowe** [reply]
This requires getting information from several sources.
First, the documentation for the XamlBinaryWriter runtime class ([UWP version](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.markup.xamlbinarywriter?view=winrt-22621), [Windows App SDK version](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.markup.xamlbinarywriter?view=windows-app-sdk-1.2)). This documentation does state that the .xbf file is a binary coded version of the plain text .xaml file. It also does state that the content of the .xaml file is the same as the .xbf file.
Second, there is the generated code for InitializeComponent.
```
void InitializeComponent()
{
    if (_contentLoaded)
        return;
           
    _contentLoaded = true;

    ::winrt::Windows::Foundation::Uri resourceLocator{ L"ms-appx:///App.xaml" };
    ::winrt::Microsoft::UI::Xaml::Application::LoadComponent(*this, resourceLocator);
}
```
```
template <typename D, typename ... I>
void MainWindowT<D, I...>::InitializeComponent()
{
    if (!_contentLoaded)
    {
        _contentLoaded = true;
        ::winrt::Windows::Foundation::Uri resourceLocator{ L"ms-appx:///MainWindow.xaml" };
        ::winrt::Microsoft::UI::Xaml::Application::LoadComponent(*this, resourceLocator, ComponentResourceLocation::Application);
    }
}
```
It only references the .xaml file, but as you know, the .xaml file isn't part of the final layout. Naturally, this would always fail if LoadComponent wasn't taking the substitution of .xbf for .xaml into account. This would hint that .xbf is substituted for the .xaml file.
So, third would obviously be the Application.LoadComponent static member ([UWP version](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.application.loadcomponent?view=winrt-22621#windows-ui-xaml-application-loadcomponent(system-object-windows-foundation-uri)), [Windows App SDK version](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.application.loadcomponent?view=windows-app-sdk-1.2#microsoft-ui-xaml-application-loadcomponent(system-object-windows-foundation-uri))). Unsurprisingly, this states that it loads a Xaml file. But it also states that this is for infrastructure only, and applications should use XamlReader.Load. While that doesn't add to much extra, what it does do is add the major implication that Application.LoadComponent is functionally similar to XamlReader.Load.
So, what does this mean?
1) The documentation outright states the equivalence of .xaml and .xbf.
2) The only way that the default layout of these projects can work is if the runtime is substituting .xbf for .xaml. LoadComponent does use .xaml in the path, but it only ever searches for .xbf. This is verifiable with ProcessMonitor.
3) The link between Application.LoadComponent and XamlReader.Load indicates similar functionality.

Also, have you actually opened up a .xbf file and compared it to the .xaml file? It is interesting.
Anyway, how does this relate to what happens in the application? Well, let's take the default MainWindow.xaml file.
```
<Window
    x:Class="Meh.MainWindow"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:Meh"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d">

    <StackPanel Orientation="Horizontal" HorizontalAlignment="Center" VerticalAlignment="Center">
        <Button x:Name="myButton" Click="myButton_Click">Click Me</Button>
    </StackPanel>
</Window>
```
The Window relates this to a Microsoft.UI.Xaml.Window runtime class. So what does it do after that?

```
MainWindow::MainWindow()
{
    auto sp = Controls::StackPanel();
    sp.Orientation(Controls::Orientation::Horizontal);
    sp.HorizontalAlignment(HorizontalAlignment::Center);
    sp.VerticalAlignment(VerticalAlignment::Center);

    auto bt = Controls::Button();
    bt.Name(L"myButton");
    bt.Click({ this, MainWindow::myButton_Click });
    bt.Content(winrt::box_value(L"Click Me"));
    sp.Children().Append(bt);
    Content(sp);
}
```
That feels right.

The .xaml files and the code behind should be seen as two parallel entities. The code behind is the behaviour of the content. The .xaml file is a means of providing build instructions for the content itself. Hence, you can infer how the runtime will initialise things. They are separately compiled entities that need to line up just right. The generated code serves as a link between the two portions.

Finally, for added context. I would like to bring attention to an older technology that does work in a similar way. The classic Windows API dialog resources. 
The main difference between these technologies is that the resource compiler converted the resource script statements to DLGTEMPLATE, DLGTEMPLATEEX, DLGITEMTEMPLATE or DLGITEMTEMPLATEEX structures. The Xaml environment just stores a preparsed binary coded version of the .xaml file.

**Reply by idexus** [reply]
@chingucoding Unlike the classes derived from FrameworkTemplate, when it comes to creating styles in the code, I managed most of it. I even managed to create styles using fluent methods, or call code while applying styles (Most of the code is generated using source generators)

<details>
<summary>example of creating style in code from my lib</summary>

```cs
var res = new ResourceDictionary() 
{
    new ThemeValue<Color> { Key = "BackgroundColor", Light = Colors.Cyan, Dark = Colors.Black },

    new Style<Button>(e => e
        .FontSize(28)
        .Width(200)
        .Height(100)                   
        .HorizontalAlignment(HorizontalAlignment.Center)),

    new Style<TextBlock>(e => e
        .HorizontalTextAlignment(TextAlignment.Center)
        .Padding(0,0,0,10)
        .FontSize(28)),

    new Style<StackPanel>(e => e
        .Padding(20))
    {
        // invoke code on StackPanel element using Style<T>
        vstack => 
        {
            vstack.Background(e => e.ResourceKey("BackgroundColor").Source(vstack));
        }
    }
};
```
</details>

What I care about is getting the ability to create applications fully in C# code, without having to resort to XAML, which in my opinion ties a man's hands in many ways. The code gives you much more possibilities.

<details>
<summary>custom control example:</summary>


```cs
// Dependency properties are generated.

[DependencyProperties]
public interface IExample
{
    public string Title { get; set; }
    public bool IsExpanded { get; set; }

    public UIElement ExampleContent { get; set; }

    [PropertyCallbacks(nameof(Example.SourceTextChanged))]
    public string SourceText { get; set; }
}

[MarkupObject]
[ContainerProperty(nameof(ExampleContent))] 
public partial class Example : Frame, IExample
{
    private readonly RichTextBlock sourceTextBlock;        

    public Example()
    {
        this.Resources.MergedDictionaries.Add(new()
        {
            new ThemeValue<Color> { Key = "BackgroundColor", Light = Colors.LightBlue, Dark = Colors.MidnightBlue }
        });

        Content = new VStack(e => e.Padding(10))
        {
            new TextBlock()
                .Text(e => e.Path(nameof(Title)).Source(this))
                .FontSize(16)
                .Margin(0,10,0,7),

            new Grid
            {
                new Frame()
                    .Content(e => e.Path(nameof(ExampleContent))
                    .Source(this))
            }
            .Margin(1, 0, 1, 0)
            .Padding(20)
            .Background(e => e.ResourceKey("BackgroundColor").Source(this)),                
                
            new Expander
            {
                new Grid
                {
                    new RichTextBlock()
                        .Assign(out sourceTextBlock)
                        .FontFamily(new FontFamily("Consolas"))
                }
                .Padding(10, 0, 10, 0)
                .Width(int.MaxValue)
            }
            .IsExpanded(e => e.Path(nameof(IsExpanded)).Source(this))
            .HorizontalAlignment(HorizontalAlignment.Stretch)
            .ExpandDirection(ExpandDirection.Down)
            .Header("Source code")
        };
    }

    public static void SourceTextChanged(object sender, DependencyPropertyChangedEventArgs e)
    {
        var example = (Example)sender;
        var sourceText = e.NewValue as string;
        var formatter = new RichTextBlockFormatter(ElementTheme.Dark);
        formatter.FormatRichTextBlock(sourceText, Languages.CSharp, example.sourceTextBlock);
    }
}
```

</details>

**Reply by marcelwgn** [reply]
Thank you for your very detailed and lengthy explanation on the difference and context of XAML to XBF and for correcting me on the matter what XBF. However this does not answer the question of how to initialize certain XAML elements such as ControlTemplate or DataTemplate in C# does it? After all, you said XAML/XBF is a way to tell the runtime how to initialize things, the question asked though was how to not use XAML to initialize everything but use pure C#.

@idexus I'm surprised that creating styles works that easily, really cool seeing that project! However, you still haven't found a solution for *Template constructs such as DataTemplate right? I think there really isn't a solution for this, during the past few years contributing to this project, I think this topic came up once or twice and everytime we resorted to creating templates using XAMLReader.Load.

----

@DarranRowe On a side note, while this is my purely my opinion, I think filling your explanations with words/sentences like "easily" and "we know" and similar constructs might be seen as arrogant or condescending.

**Reply by DarranRowe** [reply]
@chingucoding 
I wonder if you missed the biggest point of the posts. By seeing what the Xaml compiler does, you can do it yourself. You can query the object tree at runtime. You can query the controls at runtime. Why is it problematic to write an application that lays out the xaml in such a way to achieve what you want, and then query the properties of the controls?
So to directly answer the question, you use a running instance of the object tree to provide answers.

Also, for that side note. I do not see "easily" or "we know" in my post. The only "obviously" I have is in relation to Application.LoadComponent, this is backed up by the fact that I listed Xaml compiler generated code that directly references this. So I do not see why using this when I have been laying out the steps that I took to get there is that problematic. Unless you are referring to the "as you know" in relation to the .xaml files not being part of the final layout. In which case I will state that you referenced this yourself in your second reply.

**Reply by marcelwgn** [reply]
@DarranRowe Quite honestly, I think this is not a solution here. What actually does the XAML Compiler do?
Does it create XBF Files? Then it does not help with creating a DataTemplate purely using C#. 
Does the XAML Compiler generate C++ or C# Files that we should inspect? In that case, you already said, it just glues together the XAML World and the non XAML world which does not help because some stuff is created using XBF (at least I can't find the code that would create the control in the generated files beside Application.LoadComponent).
Lastly, taking a running instance of the object tree doesn't help you how it was instantiated does it? After all, @idexus and I would love to know what C# code we can use to create a DataTemplate without having to resort on any XAML Compiler during runtime.

So to conclude my view of your solutions, I am unable to find the solution to the problem, that is how to create a DataTemplate in C# without using the XAML Compiler Infrastructure during runtime given the comments you gave. I was unable to find that code in the generated files of a project and I was also unable to deduce this from a running app by inspecting the tree. But maybe I am missing some crucial detail here that would help find the solution to the question asked?

----
Regarding the side note, while you are right you did not say "we know", you did say "easy" in one of your earlier replies and "obviously" also is in the same lane of words but that's nitpicking and beside the point. It is your choice what to do with this feedback, if you think this was not righteous of me to point that out, then that's also fine. I sincerely hope that I did not offend you with these comments.

**Reply by idexus** [reply]
I am similarly unable to find a solution to the problem based on what @DarranRowe writes. The main problem is that when you define a *Template, you have no idea what's in it until you instantiate it. 

Here we need a solution to this fairly simple problem directly in the library, not some XAML workaround. I know the Microsoft world is the XAML world these days, but not everyone wants to build applications using this technology. Maybe someone who decides will hear it.

**Reply by DarranRowe** [reply]
I see, that is unfortunate.
I will have to bow out then. This would require a lot more work than I am willing to put in on my rest days, and I am currently too busy during the week to be able to put in the required work.
I hope that someone else gives you a satisfactory answer.

There is also one major thing that I feel needs to be corrected. This was never about providing Xaml workarounds. The fact that things were seen this way shows that I was not the right person. This had a much simpler goal. If you understand how the Xaml runtime does things, then it means that you understand the basic concepts. The result of this is that you then understand how to write the exact same thing in code. Figuring things out was only ever a means to an end.
